bin/exec-env

```diff
@@ -1,36 +1,57 @@
-#!/usr/bin/env bash
+#! /usr/bin/env bash
 
 set -euo pipefail
 
+pyproject=${MISE_TOOL_OPTS__PYPROJECT-}
+if [[ -n ${MISE_PROJECT_ROOT-} && -n ${pyproject} && ${pyproject} != /* ]]; then
+  pyproject="${MISE_PROJECT_ROOT}/${pyproject}"
+fi
+
 # shellcheck source-path=.
-source "$MISE_PLUGIN_PATH/bin/utils.sh"
+source "${MISE_PLUGIN_PATH}/bin/utils.sh"
 
 setup_virtualenv() {
-  # if `uv` is being used, exit early and let the user manage venv activation
-  if [[ -f "uv.lock" ]]; then
+  if [[ -z ${pyproject} ]]; then
+    # tool is installed (maybe globally) but project is not activated
+    return
+  elif [[ ! -f ${pyproject} ]]; then
+    echoerr "Project file not found (${pyproject}). Execute \`poetry init\` to create it."
     return
   fi
 
-  set +e
-  venv="$(poetry_venv)"
-  poetry_venv_exit_code=$?
-  set -e
+  local project_root=${pyproject%/*}
 
-  if [[ $poetry_venv_exit_code -eq 1 ]]; then
+  # if `uv` is being used, exit early and let the user manage venv activation
+  if [[ -f "${project_root}/uv.lock" ]]; then
     return
   fi
 
-  if [[ ! -d "$venv" ]]; then
-    if [[ "${MISE_POETRY_AUTO_INSTALL:-}" != "true" ]] && [[ "${MISE_POETRY_AUTO_INSTALL:-}" != "1" ]]; then
+  # don't activate environment if lock file isn't present
+  if is_enabled MISE_POETRY_VENV_AUTO; then
+    if [[ ! -f "${project_root}/poetry.lock" ]]; then
       return
     fi
-    poetry_bin install
-    venv="$(poetry_bin env info --path)"
   fi
 
-  POETRY_ACTIVE=1
-  MISE_ADD_PATH="$venv/bin"
-  export VIRTUAL_ENV="$venv" POETRY_ACTIVE MISE_ADD_PATH
+  local virtual_env
+  if ! virtual_env=$(poetry_venv "${project_root}"); then
+    # run a dummy command to ensure the venv exists (creating it if possible)
+    poetry_bin "${project_root}" --ansi run true
+
+    if ! virtual_env=$(poetry_venv "${project_root}"); then
+      echoerr "Unable to create/activate virtual environment."
+      return
+    fi
+
+    if is_enabled MISE_POETRY_AUTO_INSTALL; then
+      echoerr "Executing \`poetry install\` to populate virtual environment."
+      poetry_bin "${project_root}" --ansi install
```

Comment:
there's `MISE_COLOR` but I'm not sure it gets passed into plugins - is that the one you were thinking of?
There is also reference to `CLICOLOR` in the docs, but I don't think that gets passed in either.

`poetry` uses the `NO_COLOR` env var for the same purpose, I think.

The reason I had set `--ansi` because when `poetry` runs non-interactively in the plugin it doesn't display colours.

Reply:
there are situations like if `TERM=dumb` where it can be a tty but colors should still not be enabled. I think we should add a new env var like `MISE_PLUGIN_COLORS=1` to this list: https://github.com/jdx/mise/blob/ba66d4659c6d8f3ffa589dacfe402d6988e46d9a/src/plugins/asdf_plugin.rs#L410-L416

we don't want to use `MISE_COLOR` I don't think because mise might be used for non-tty things inside of a plugin

Reply:
OK - I've added a check on (the currently non-existent) `MISE_PLUGIN_COLORS` which means that `--ansi` is disabled for now